kernel/kernel-default/src/test/java/io/delta/kernel/client/TestDefaultJsonHandler.java

```diff
@@ -0,0 +1,181 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.client;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import static io.delta.kernel.utils.DefaultKernelTestUtils.getTestResourceFilePath;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.hadoop.conf.Configuration;
+import static org.junit.Assert.assertEquals;
+import org.junit.Test;
+
+import io.delta.kernel.data.ColumnarBatch;
+import io.delta.kernel.data.DefaultJsonRow;
+import io.delta.kernel.data.FileDataReadResult;
+import io.delta.kernel.data.Row;
+import io.delta.kernel.expressions.Literal;
+import io.delta.kernel.fs.FileStatus;
+import io.delta.kernel.types.BooleanType;
+import io.delta.kernel.types.LongType;
+import io.delta.kernel.types.MapType;
+import io.delta.kernel.types.StringType;
+import io.delta.kernel.types.StructType;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.Utils;
+
+public class TestDefaultJsonHandler
+{
+    private static final JsonHandler JSON_HANDLER = new DefaultJsonHandler(new Configuration());
+    private static final FileSystemClient FS_CLIENT =
+        new DefaultFileSystemClient(new Configuration());
+
+    @Test
+    public void contextualizeFiles()
+        throws Exception
+    {
+        try (CloseableIterator<Row> inputScanFiles = testFiles();
+            CloseableIterator<FileReadContext> fileReadContexts =
+                JSON_HANDLER.contextualizeFileReads(testFiles(), Literal.TRUE)) {
+            while (inputScanFiles.hasNext() || fileReadContexts.hasNext()) {
+                assertEquals(inputScanFiles.hasNext(), fileReadContexts.hasNext());
+                Row inputScanFile = inputScanFiles.next();
+                FileReadContext outputScanContext = fileReadContexts.next();
+                compareScanFileRows(inputScanFile, outputScanContext.getScanFileRow());
+            }
+        }
+    }
+
+    @Test
+    public void readJsonFiles()
+        throws Exception
+    {
+        CloseableIterator<FileDataReadResult> data =
+            JSON_HANDLER.readJsonFiles(
+                JSON_HANDLER.contextualizeFileReads(testFiles(), Literal.TRUE),
+                new StructType()
+                    .add("path", StringType.INSTANCE)
+                    .add("size", LongType.INSTANCE)
+                    .add("dataChange", BooleanType.INSTANCE)
+            );
+
+        List<String> actPaths = new ArrayList<>();
+        List<Long> actSizes = new ArrayList<>();
+        List<Boolean> actDataChanges = new ArrayList<>();
+        while (data.hasNext()) {
```

Comment:
Can we call `hasNext` an extra time here before to be sure it's idempotent? And run this test with batchSize = 1?

Reply:
done